internal/gather/gather_bytes.go

```diff
@@ -120,6 +124,78 @@ type bytesReadSeekCloser struct {
 	offset int
 }
 
+func (b *bytesReadSeekCloser) ReadAt(bs []byte, off int64) (int, error) {
+	b.b.assertValid()
+	// cache "b.b.Slices" - slice parameters will stay constant for duration of
+	// function.  Locking is left to the calling function
+	slices := b.b.Slices
+
+	// source data that is read will be written to w, the buffer backed by p.
+	offset := off
+
+	maxBsIndex := len(bs)
+
+	// negative offsets result in an error
+	if offset < 0 {
+		return 0, ErrInvalidOffset
+	}
+
+	sliceNdx := -1
+
+	// find the index of starting slice
+	for i, slicesBuf := range slices {
+		if offset < int64(len(slicesBuf)) {
+			sliceNdx = i
+			break
+		}
+
+		// update offset to be relative to the sliceNdx slice
+		offset -= int64(len(slicesBuf))
+	}
+
+	// no slice found if sliceNdx is still negative
+	if sliceNdx == -1 {
+		// return no bytes read if the buffer has no length
+		if maxBsIndex == 0 {
+			return 0, nil
+		}
+
+		return 0, io.EOF
+	}
+
+	curSlice := slices[sliceNdx]
+
+	m := copy(bs, curSlice[offset:])
+	// accounting: keep track of total number of bytes written in n and
+	// number of bytes written from the current slice in m
+	n := m
+
+	// move on to next and then check if all slices were consumed
+	sliceNdx++
+
+	slicesN := len(slices)
+
+	for maxBsIndex > n && sliceNdx < slicesN {
```

Comment:
what happens if `maxBsIndex = n` ? We probably do not want to enter this loop then, right ?

Reply:
I think the cop() below is probably saving you from a panic here, but better to not even enter this loop when we have filled the input buffer